#2: added initial project script, .gitignore and README.md files (#3)
exasol/row-level-security: doc/system_requirements.md

<head><link href="oft_spec.css" rel="stylesheet"></link></head>

> # System Requirement Specification Exasol Row / Column Level Security

Author: ```suggestion
# System Requirement Specification Exasol Row Level Security
```

## Introduction

> Exasol Row / Column Level Security (short "RCLS") is a plug-in for Exasol that provides the ability to grant access on individual rows or columns based on contents of the database.

Author: ```suggestion
Exasol Row Level Security (short "RLS") is a plug-in for Exasol that provides the ability to grant access on individual rows based on contents of the database.
```

## About This Document

### Target Audience

The target audience are end-users, requirement engineers, software designers and quality assurance. See section ["Stakeholders"](#stakeholders) for more details.

### Goal

> The RCLS main goal is to provide fine-grained access control below the level of tables.

Author: ```suggestion
The RLS main goal is to provide fine-grained access control below the level of tables.
```

### Quality Goals

> RCLS's main quality goals are in descending order of importance:

Author: ```suggestion
RLS's main quality goals are in descending order of importance:
```

1. Provide reliable security
1. Work with an affordable performance hit compared to accessing the data without RCLS

## Stakeholders

### Data Owners

TODO

### Regular Users

TODO

### Terms and Abbreviations

The following list gives you an overview of terms and abbreviations commonly used in OFT documents.

* Column: An Attribute of a defined type shared by all datasets in a table.
* Row: Dataset in the database

In the following subsections central terms are explained in more detail.

## Features

> Features are the highest level requirements in this document that describe the main functionality of RCLS.

Author: ```suggestion
Features are the highest level requirements in this document that describe the main functionality of RLS.
```

### Row Level Security
`feat~row-level-security~1`

> RCLS lets administrators grant access to individual table rows based on configurable criteria.

Author: ```suggestion
RLS lets administrators grant access to individual table rows based on configurable criteria.
```

Needs: req

## Functional Requirements

### Common Requirements

> In this section you find requirements that are common for both row- and column-level security. 

Author: Do we even need a common requirements section if we only have row level security?

Reviewer: No, we don't.

Author: removed it

#### User Roles
`req~user-roles~1`

Data Owners can assign zero or more roles to users.

Rationale:

> Roles are used to determine whether a user may or may not access data in RCLS.

Author: ```suggestion
Roles are used to determine whether a user may or may not access data in RLS.
```

#### Granting Row Read Access Based on User Role
`req~granting-row-read-access-based-on-user-role~1`

Data Owners can define the role a user must have in order to read a row.

Needs: dsn

#### Public Rows
`req~public-rows~1`

Data Owners can define rows that all users can read, regardless of those users' roles.

Rationale:

This allows data owners to make non-confidential data public in a table otherwise restricted by row-level security.

Covers:

* [feat~row-level-security~1](#row-level-security)

Needs: dsn

#### Rows Users are Allowed to Read
`req~rows-users-are-allowed-to-read~1`

A user can read a row if at least one of the roles assigned to the user is also assigned to that row.

Rationale:

We want rows to be potentially readable by multiple roles. If one of the user roles matches that is sufficient. Having multiple roles match means overachieving the criteria. In a role-based security model it also makes no sense to require someone having more than one role to access a row. 

Covers:

* [feat~row-level-security~1](#row-level-security)

Needs: dsn

#### Only Accessible Rows Contribute to Aggregate Functions
`req~only-accessible-rows-contribute-to-aggregate-functions~1`

> If row access is controlled by RCLS and a user invokes an aggregate function, then only the rows accessible to that user contributed to the result of the aggregate function.

Author: ```suggestion
If row access is controlled by RLS and a user invokes an aggregate function, then only the rows accessible to that user contributed to the result of the aggregate function.
```

<head><link href="oft_spec.css" rel="stylesheet"></link></head>

# System Requirement Specification Exasol Row Level Security

## Introduction

Exasol Row Level Security (short "RLS") is a plug-in for Exasol that provides the ability to grant access on individual rows based on contents of the database.

## About This Document

### Target Audience

The target audience are end-users, requirement engineers, software designers and quality assurance. See section ["Stakeholders"](#stakeholders) for more details.

### Goal

The RLS main goal is to provide fine-grained access control below the level of tables.

### Quality Goals

RLS's main quality goals are in descending order of importance:

1. Provide reliable security
1. Work with an affordable performance hit compared to accessing the data without RCLS

## Stakeholders

### Data Owners

> TODO

Reviewer: ```suggestion
Data Owners are have full access to the data _before_ row-level security is applied. They also get to decide who is allowed to read it through RLS.
```

Author: done

### Regular Users

> TODO

Reviewer: ```suggestion
Regular Users in the context of this project are consumers of row-level security-protected data.
```

Author: done
